### PR TITLE
fix(snap): Install Unit of Measure config file on upgrade

### PR DESCRIPTION
**Pull request by farshidtz**
PR #4119 added the following environment variable:
https://github.com/edgexfoundry/edgex-go/blob/166620d9f56c7df6403b4f88f57b6f2fdd40f3fd/snap/snapcraft.yaml#L278

This makes the edge releases of snap backward incompatible as of v2.3.0-dev.45 because this file doesn't exist when upgrading from an older version. The latest snap beta release which isn't broken is 2.3.0-dev.44 (snap rev 5898/5899). This fix works when upgrading from the latest beta or older.

This PR adds the logic to add the configuration file when upgrading (i.e. using `snap refresh`).

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions

Reproduce the error:
```
$ snap install edgexfoundry --channel=latest/stable
edgexfoundry 2.2.0+2 from Canonical✓ installed

$ snap refresh edgexfoundry --channel=latest/edge
edgexfoundry (edge) 2.3.0-dev.46 from Canonical✓ refreshed

$ snap logs -n 100 edgexfoundry.core-metadata | grep -i ERROR
2022-08-15T16:16:26+02:00 edgexfoundry.core-metadata[386550]: level=ERROR ts=2022-08-15T14:16:26.714636738Z app=core-metadata source=httpserver.go:143 msg="Web server failed: http: Server closed"
2022-08-15T16:17:40+02:00 edgexfoundry.core-metadata[388065]: level=ERROR ts=2022-08-15T14:17:40.24683483Z app=core-metadata source=bootstrap.go:42 msg="could not load unit of measure configuration file (/var/snap/edgexfoundry/3904/config/core-metadata/res/uom.toml): open /var/snap/edgexfoundry/3904/config/core-metadata/res/uom.toml: no such file or directory"
```

Test the changes using the snap built from this branch:
```
$ snap install edgexfoundry --channel=latest/stable
edgexfoundry 2.2.0+2 from Canonical✓ installed

$ snap refresh edgexfoundry --channel=latest/edge/pr-4125
edgexfoundry (edge/pr-4125) 2.3.0-dev.46 from Canonical✓ refreshed

$ snap logs -n 2 edgexfoundry.core-metadata 
2022-08-15T16:25:57+02:00 edgexfoundry.core-metadata[395550]: level=INFO ts=2022-08-15T14:25:57.308622394Z app=core-metadata source=message.go:55 msg="This is the EdgeX Core Metadata Microservice"
2022-08-15T16:25:57+02:00 edgexfoundry.core-metadata[395550]: level=INFO ts=2022-08-15T14:25:57.308628766Z app=core-metadata source=message.go:58 msg="Service started in: 85.428935ms"

$ sudo journalctl -n 100 | grep post-refresh
systemd[1]: Started snap.edgexfoundry.hook.post-refresh.0409b2f7-0d0b-42dd-97e5-e865eaf1dece.scope.
edgexfoundry.post-refresh[394831]: Refreshing from  (3722) to 2.3.0-dev.46 (3912)
...
edgexfoundry.post-refresh[394843]: Installing /snap/edgexfoundry/3912/config/core-metadata/res/uom.toml
systemd[1]: snap.edgexfoundry.hook.post-refresh.0409b2f7-0d0b-42dd-97e5-e865eaf1dece.scope: Deactivated successfully.
```

### CODE REVIEW

File: snap/hooks/pre-refresh [hunk 1]
```diff
@@ -1,11 +1,10 @@
 #!/bin/bash -e
 
-# save this revision for when we run again in the post-refresh
+# save these for use in the post-refresh hook
 snapctl set lastrev="$SNAP_REVISION"
-
-snapctl set release="jakarta"
+snapctl set lastver=$SNAP_VERSION
```
Comment (farshidtz) [inline snap/hooks/pre-refresh:5]: Replace outdated `release` with the problematically set `lastver`. This isn't needed for this fix, but a technical dept worth getting rid of. The value of `lastver` could be used in future to implement migration logic between major releases.

Comment (farshidtz) [reply snap/hooks/pre-refresh]: For clarity outside of the refresh hooks context, I've refactored as bellow:
```
lastrev -> pre-refresh.revision
lastver -> pre-refresh.version
```

The options will then be store as:
```
$ snap get edgexfoundry pre-refresh -d
{
	"pre-refresh": {
		"revision": 3915,
		"version": "2.3.0-dev.46"
	}
}
```

I also made sure we continue reading the old `lastrev` to handle the config file installation when refreshing from older revisions.

File: snap/hooks/pre-refresh [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
-snapctl set lastrev="$SNAP_REVISION"
+# save these for use in the post-refresh hook
+snapctl set pre-refresh.revision=$SNAP_REVISION
+snapctl set pre-refresh.version=$SNAP_VERSION
+
+# unset legacy EdgeX <2.3 options
+snapctl unset lastrev release
 
-snapctl set release="jakarta"
-
```
Comment (farshidtz) [inline snap/hooks/pre-refresh:7]: @MonicaisHer corrected the snap option name.

**Review by MonicaisHer — APPROVED** [review]
@farshidtz Thank you, looks good!

**Comment by farshidtz** [comment]
@cloudxxx8 could you please approve this? It's been reviewed.

**Review by lenny-intel — APPROVED** [review]

**Pull request merged**